ably/realtime_conn_spec_test.go

```diff
@@ -121,8 +123,147 @@ func (c protoConnWithFakeEOF) Receive(deadline time.Time) (*proto.ProtocolMessag
 
 	select {
 	case r := <-received:
+		if c.onMessage != nil {
+			c.onMessage(r.msg)
+		}
 		return r.msg, r.err
 	case <-c.doEOF:
 		return nil, io.EOF
 	}
 }
+
+func TestRealtimeConn_RTN15b(t *testing.T) {
+	t.Parallel()
+
+	doEOF := make(chan struct{}, 1)
+
+	type meta struct {
+		dial     *url.URL
+		messages []*proto.ProtocolMessage
+	}
+
+	var metaList []*meta
+
+	app, client := ablytest.NewRealtimeClient(&ably.ClientOptions{
+		NoConnect: true,
+		Dial: func(protocol string, u *url.URL) (proto.Conn, error) {
+			m := &meta{dial: u}
+			metaList = append(metaList, m)
+			c, err := ablyutil.DialWebsocket(protocol, u)
+			return protoConnWithFakeEOF{Conn: c, doEOF: doEOF, onMessage: func(msg *proto.ProtocolMessage) {
+				m.messages = append(m.messages, msg)
+			}}, err
+		},
+	})
+	defer safeclose(t, client, app)
+
+	if err := ablytest.Wait(client.Connection.Connect()); err != nil {
+		t.Fatalf("Connect=%s", err)
+	}
+
+	channel := client.Channels.Get("channel")
+
+	if err := ablytest.Wait(channel.Attach()); err != nil {
+		t.Fatal(err)
+	}
+
+	sub, err := channel.Subscribe()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	stateChanges := make(chan ably.State, 16)
+	client.Connection.On(stateChanges)
+
+	doEOF <- struct{}{}
+
+	var state ably.State
+
+	select {
+	case state = <-stateChanges:
+	case <-time.After(50 * time.Millisecond):
+		t.Fatal("didn't transition on EOF")
+	}
+
+	if expected, got := ably.StateConnDisconnected, state; expected != got.State {
+		t.Fatalf("expected transition to %v, got %v", expected, got)
+	}
+
+	// Publish a message to the channel through REST. If connection recovery
+	// succeeds, we should then receive it without reattaching.
+
+	rest, err := ably.NewRestClient(app.Options())
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = rest.Channels.Get("channel", nil).Publish("name", "data")
```

Comment:
There's a race here between this publish and the reconnection. It could happen that, before this publish, the connection already reconnected, in which case we wouldn't know if we recovered the connection state or not.

I'd suggest putting a channel send/rececive in the `Dial` function above and do this:

- Wait for Dial to be called; let it blocked there.
- Do the publish.
- Let Dial to finish connection.

Kind of like:

```go
gotDial := make(chan chan struct{})

Dial: func ... {
    ...
    goOn := make(chan struct{})
    gotDial <- goOn
    <-goOn
   ....
}

goOn := <-gotDial
rest.Publish(...)
close(goOn)
```

Reply:
addressed e0b5de36d73a7913fea09ab1e190ce32838f3f9b